### PR TITLE
Dynamic event name proposal

### PR DESCRIPTION
**Pull request by mheins57**

### CODE REVIEW

File: proposals/0000-allow-dynamic-event-names.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
+---
+title: Allowing Dynamic Event Names in RCTEventEmitter
+author:
+- Michael Heins
+date: today
```
[the remark this answers is not in the document]
Comment (mheins57) [reply proposals/0000-allow-dynamic-event-names.md]: Not sure how to add a comment to my original pull request?

**Comment by axemclion** [comment]
@mheins57 - Thanks for the proposal ? 

Could you also add some more details on how this would be done ? 

cc @ shergin

**Comment by TheSavior** [comment]
Something to think about with this is how these events could be typed with Flow / TypeScript to ensure that the values people receive are the correct shape and that they don't mistype the event names. 

I'm not sure how we could do this with dynamic event names.

**Comment by mheins57** [comment]
As I mentioned in the Detailed Design section of the ReadMe, I propose to add a -supportsEventName:(NSString *)eventName method to RCTEventEmitter, and then modify two methods: -sendEventWithName:body:  and -addListener: to call the supportsEventName: method.  If that returns NO, the older supportedEvents method would be checked, as before.  I think those are the only changes required.  In this way, no existing subclasses of RCTEventEmitter would be modified, and the checking would be exactly as it was before.  A new subclass of RCTEventEmitter wishing to use dynamic names would override the default -supportsEventName: method to provide custom checking of the event name.  

For example, in the use cases I had in mind, event names would consist of a two part string.  The first part of the string could be a fixed portion, such as the module name, for instance "RNBluetoothLowEnergy-".  The variable part could be a UUID taken from a discovered wireless device.  The event name would be the concatenation of these two parts.  To do the checking, the overridden supportsEventName: method could reject any event name not starting with the module name, and it could also apply some sort of format check for the variable UUID part, or even keep track of any relevant device UUIDs that it had discovered and use that for comparison -- whatever is appropriate for the module design.

**Comment by mheins57** [comment]
What's the next step?

**Comment by cpojer** [comment]
I'm fine with this addition and not too concerned with typing here. I also don't think we need to merge this PR to keep the RFC for eternity as it is a rather small change. Please just send a PR to react-native with your proposed change and we'll handle it from there.

**Comment by mheins57** [comment]
Thank you

Michael Heins

Any fool can know. The point is to understand. — A. Einstein

**Pull request closed without merging**